src/v/cloud_storage/remote_partition.h

```diff
@@ -331,6 +332,12 @@ class remote_partition : public ss::enable_shared_from_this<remote_partition> {
       materialized_segment_state,
       &materialized_segment_state::_hook>
       _materialized;
+
+    // Limit the number of readers of materialized segments that may exist.
+    // This also implicitly limits the number of materialized segments that may
+    // exist, since each materialized segment should have at least one reader.
+    ssx::semaphore _readers_limit;
```

Comment:
Maybe it will be better to have this limit per-shard instead of per-partition? Some partitions may be accessed more frequently then the others.